demosplan/DemosPlanCoreBundle/ValueObject/BasicKeycloakUserData.php

```diff
@@ -0,0 +1,150 @@
+     */
+    protected string $organisationName = '';
+
+    /**
+     * Unique identifier of the provided organisation.
+     */
+    protected string $organisationId = '';
+
+    protected string $firstName = '';
+
+    protected string $lastName = '';
+
+    public function fill(ResourceOwnerInterface $resourceOwner): void
+    {
+        $userInformation = $resourceOwner->toArray();
+
+        if (array_key_exists('groups', $userInformation)
+            && is_array($userInformation['groups'])
+        ) {
+            $this->mapCustomerRoles($userInformation['groups']);
+        }
+
+        $this->userId = $userInformation['sub'] ?? '';
+        $this->organisationName = $userInformation['organisationName'] ?? '';
+        $this->organisationId = $userInformation['organisationId'] ?? '';
+        $this->firstName = $userInformation['given_name'] ?? '';
+        $this->lastName = $userInformation['family_name'] ?? '';
+        $this->userName = $userInformation['preferred_username'] ?? ''; // kind of "login" //has to be unique?
+        $this->emailAddress = $userInformation['email'] ?? '';
+
+        $this->lock();
+        $this->checkMandatoryValuesExist();
+    }
+
+    /**
+     * Checks for existing mandatory data.
+     */
+    public function checkMandatoryValuesExist(): void
+    {
+        $missingMandatoryValues = [];
+        if ('' === $this->userId) {
+            $missingMandatoryValues[] = 'userId';
+        }
+
+        if ('' === $this->userName) {
+            $missingMandatoryValues[] = 'userName';
+        }
+
+        if ('' === $this->emailAddress) {
+            $missingMandatoryValues[] = 'emailAddress';
+        }
+
+        if ('' === $this->organisationId) {
+            $missingMandatoryValues[] = 'organisationId';
+        }
+
+        if ('' === $this->firstName && '' === $this->lastName) {
+            $missingMandatoryValues[] = 'name';
+        }
+
+        if ([] === $this->customerRoleRelations) {
+            $missingMandatoryValues[] = 'roles';
+        }
+
+        if ([] !== $missingMandatoryValues) {
+            throw new AuthenticationCredentialsNotFoundException(implode(', ', $missingMandatoryValues).'are missing in requestValues');
+        }
+    }
+
+    /**
+     * @param array<int, string> $groups
+     */
+    private function mapCustomerRoles(mixed $groups): void
+    {
+        foreach ($groups as $group) {
+            $subGroups = explode('/', $group);
+            if (str_contains($subGroups[1], 'Beteiligung-Berechtigung')) {
+                $subdomain = strtolower(explode('-', $subGroups[2])[0]);
+                $this->customerRoleRelations[$subdomain][] = $subGroups[3];
+            }
+        }
+    }
+
+    public function __toString(): string
+    {
+        $customerRoleRelationString = '-';
+        foreach ($this->customerRoleRelations as $subdomain => $roleNames) {
+            $customerRoleRelationString = $subdomain.': ['.implode(', ', $roleNames).']';
```

Comment:
Do you intend to overwrite $customerRoleRelationString on every loop iteration?

Reply:
No, thanks.
d1b49c5d693d9a8ee0318511d056314dbc342adf